protobuf-c: initial onboarding (v1.4.1)

gegles opened the pull request:
Specify library name and version:  **protobuf-c/1.4.1**

This PR aims to onboard the [protobuf-c](https://github.com/protobuf-c/protobuf-c) lib.

I tried to follow the template as much as possible, but I also got inspiration from the `protobuf` recipe as there are some similarities.

The difficulty with this project is that, similarly to `protobuf` it provide a code generator that will be used as a build dependency, but also a library that is added as a normal dependency.

It also provides a `cmake_build_module` that is loaded inside the `find_package` to make the `protobuf_generate_c` method available inside the user's `CMakeLists.txt`..

The test_package provided tries to put all these aspects together:
* Have `protobuf-c` as both a `tool_requires` and a `requires`
* Have the `CMakeLists.txt` use the `protobuf_generate_c` method to generate code from a `.proto` file
* Have the test program use the generated code

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

gegles commented:
FYI @bemehiser

gegles commented:
@prince-chrismc, not sure if you have any quick tip for me, but I can't seem to repro the following issue on my local setup (RHEL8 docker with GCC12):
```
Unknown CMake command "protobuf_generate_c".
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=17295) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=17295) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=17295) it.</sub>

memsharded commented on recipes/protobuf-c/all/conanfile.py:
```diff
@@ -0,0 +1,101 @@
+    def package_info(self):
+        self.cpp_info.libs = ["protobuf-c"]
+
+        self.cpp_info.set_property("cmake_file_name", "protobuf-c")
+        self.cpp_info.set_property("cmake_target_name", "protobuf-c::protobuf-c")
+        self.cpp_info.set_property("pkg_config_name", "protobuf-c")
```
```suggestion
        # upstream CMake config file name and target name matches the package name
```
This is not necessary, as the package name matches the upstream. A comment to show this was checked is enough.

memsharded commented on recipes/protobuf-c/all/test_v1_package/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.15)
```
test_v1_package tests are no longer necessary, that was removed some days ago

gegles replied:
Thanks. Removed. It might be good to remove it from the template [here](https://github.com/gegles/conan-center-index/tree/master/docs/package_templates/cmake_package/all/test_v1_package) then.

memsharded commented on recipes/protobuf-c/all/conanfile.py:
```diff
@@ -0,0 +1,101 @@
+class ProtobufCConan(ConanFile):
+    name = "protobuf-c"
+    description = "Protocol Buffers implementation in C"
+    license = "LicenseRef-LICENSE"
```
Is this a valid license?

gegles commented:
@memsharded @prince-chrismc @RubenRBS @uilianries @bemehiser, this PR is finally ready to be properly reviewed... All builds are now passing. I banged my head against the wall for several hours to try and get the whole `shared` lib working, but with the double level of `protobuf` and then `protobuf-c` as both libraries and tools, it got real tricky to figure out what was exactly needed. Most folks should be OK with `static` for now and we can always improve later...

Thanks in advance for your reviews!

uilianries reviewed (commented):
@gegles Thank you for your contribution, Protobuf is a real challenge to be packaged, I see that you needed to fix important parts by using new patches. It would be great if you could send these new patches to the upstream too, as PRs.

uilianries commented on recipes/protobuf-c/all/conandata.yml:
```diff
@@ -0,0 +1,9 @@
+patches:
+  "1.4.1":
+    - patch_file: "patches/1.4.1-cmake-protobuf-target.patch"
+      patch_description: "Fixup the build"
```
```suggestion
      patch_description: "Fixup the build"
```
It does not help much, please, describe the real meaning, like "Fit Protobuf directory variables to official targets"

gegles commented on recipes/protobuf-c/all/patches/1.4.1-cmake-protobuf-target.patch:
```diff
@@ -0,0 +1,37 @@
+ ADD_CUSTOM_COMMAND(OUTPUT protobuf-c/protobuf-c.pb.cc protobuf-c/protobuf-c.pb.h
+                    COMMAND ${PROTOBUF_PROTOC_EXECUTABLE}
+-                   ARGS --cpp_out ${CMAKE_BINARY_DIR} -I${MAIN_DIR} ${MAIN_DIR}/protobuf-c/protobuf-c.proto)
++                   ARGS --cpp_out ${CMAKE_BINARY_DIR} -I${MAIN_DIR} -I${protobuf_INCLUDE_DIR} ${MAIN_DIR}/protobuf-c/protobuf-c.proto)
```
Waiting on upstream [PR](https://github.com/protobuf-c/protobuf-c/pull/555) to go through to address this.

gegles commented on recipes/protobuf-c/all/protobuf-gen-c.cmake:
```diff
@@ -0,0 +1,40 @@
+find_program(PROTOC_EXECUTABLE protoc REQUIRED)
```
Waiting on upstream https://github.com/protobuf-c/protobuf-c/pull/555 to go through to address this.

gegles commented on recipes/protobuf-c/all/conanfile.py:
```diff
@@ -0,0 +1,105 @@
+        "with_protoc": [True, False]
+    }
+    default_options = {
+        # "shared": False,
```
Ok, I've now added a #FIXME. As I mentioned earlier: 

> I banged my head against the wall for several hours to try and get the whole shared lib working, but with the double level of protobuf and then protobuf-c as both libraries and tools, it got real tricky to figure out what was exactly needed. Most folks should be OK with static for now and we can always improve later...

gegles commented:
Ok @uilianries, I've addressed your comments, pushed a [PR upstream](https://github.com/protobuf-c/protobuf-c/pull/555) and adjusted the recipe accordingly.
Given the lack of activity on the main project (lots of stale PRs), I would hope we could merge this in and hopefully in the future clean the patch out once a v1.4.2 comes around...

In the meantime, it would be super useful if this could be merged in as it is literally our last blocker to transition to Conan v2. 

Much appreciated!

uilianries commented on recipes/protobuf-c/all/conanfile.py:
```diff
@@ -0,0 +1,104 @@
+        self.cpp_info.set_property("cmake_build_modules", [
+            os.path.join(self._cmake_install_base_path, "protobuf-c.cmake")
+        ])
+        self.buildenv_info.append_path("PATH", os.path.join(self.package_folder, "bin"))
```
```suggestion
```
For Conan 2.x both buildenv_info and runenv_info are automatically added to PATH.

gegles replied:
@uilianries, if I remove this line I get the following error:
```
CMake Error at /Users/gegles/.conan2/p/b/proto04030cf11be22/p/lib/cmake/protobuf-c/protobuf-c.cmake:2 (find_program):
  Could not find PROTOC_GEN_C using the following names: protoc-gen-c
Call Stack (most recent call first):
  build/apple-clang-14-x86_64-17-release/generators/protobuf-c-config.cmake:37 (include)
  CMakeLists.txt:4 (find_package)
```

uilianries reviewed (approved):
LGTM. Thank you for submitting your patch to the upstream. This PR looks good enough now.

gegles commented:
Thanks for your review @uilianries! 

@memsharded @danimtb @jcar87, any chance you could approve/merge this soon?

Much appreciated!
